src/Native/System.Native/pal_io.cpp

```diff
@@ -202,3 +214,83 @@ int32_t ShmUnlink(const char* name)
 {
     return shm_unlink(name);
 }
+
+
+static
+void ConvertDirent(const dirent& entry, DirectoryEntry* outputEntry)
+{
+    // We use Marshal.PtrToStringAnsi on the managed side, which takes a pointer to
+    // the start of the unmanaged string. Give the caller back a pointer to the 
+    // location of the start of the string that exists in their own byte buffer.
+    outputEntry->nameStartAddress = &entry.d_name[0];
+    outputEntry->inode_type = (NodeType)entry.d_type;
+
+#if HAVE_DIRENT_NAME_LEN
+    outputEntry->nameLength = entry.d_namlen;
+#else
+    outputEntry->nameLength = -1; // sentinel value to mean we have to walk to find the first \0
+#endif
+}
+
+extern "C" 
+int32_t GetDirentSize()
+{
+    // dirent should be under 2k in size
+    static_assert(sizeof(dirent) < 2048, "");
+    return static_cast<int32_t>(sizeof(dirent));
+}
+
+// To reduce the number of string copies, this function calling pattern works as follows:
+// 1) The managed code calls GetDirentSize() to get the platform-specific 
+//    size of the dirent struct.
+// 2) The managed code creates a byte[] buffer of the size of the native dirent
+//    and passes a pointer to this buffer to this function. 
+// 3) This function passes input byte[] buffer to the OS to fill with dirent data
+//    which makes the 1st strcpy.
+// 4) The ConvertDirent function will set a pointer to the start of the inode name
+//    in the byte[] buffer so the managed code and find it and copy it out of the 
+//    buffer into a managed string that the caller of the framework can use, making
+//    the 2nd and final strcpy. 
+extern "C" 
+bool ReadDirR(DIR* dir, void* buffer, int32_t bufferSize, DirectoryEntry* outputEntry)
+{
+    assert(bufferSize == sizeof(dirent));
+    assert(buffer != nullptr);
+    assert(dir != nullptr);
+    assert(outputEntry != nullptr);
+
+    bool haveData = false;
+    errno = 0;
```

Comment:
This was for when we captured the error result since the man page says to clear errno first. Since we aren't capturing the error result anymore, that can go away